### PR TITLE
Fix #2081: Prevent infinite loop

### LINKED ISSUE #2081: Game hangs when track has infinite loop
screenshot 25

Track Merged, Hit test. Nothing else was changed.

### PR DESCRIPTION
**Pull request by jensj12**
Adds an extra iterator that steps forward at half the speed of the normal iterator. If they ever meet, there's a loop in the track and the track is classified as incomplete.
happy park sb 2016-08-31 17-49-44

### CODE REVIEW

File: src/ride/ride.c [hunk 1]
```diff
@@ -756,12 +756,24 @@ int ride_find_track_gap(rct_xy_element *input, rct_xy_element *output)
 	if (w != NULL && _rideConstructionState != RIDE_CONSTRUCTION_STATE_0 && _currentRideIndex == rideIndex)
 		sub_6C9627();
 
+	uint8 counter = 1;
```
Comment (janisozaur) [inline src/ride/ride.c]: this could be `bool`

File: distribution/changelog.txt [hunk 2]
```diff
@@ -28,6 +29,7 @@
 - Removed: BMP screenshots.
 - Removed: Intamin and Phoenix easter eggs.
 - Fix: [#1038] Guest List is out of order.
+- Fix: [#2081] Infinite loop.
```
Comment (IntelOrca) [inline distribution/changelog.txt]: I have updated the issue to something more meaningful:
`Game hangs when track has infinite loop`

File: src/ride/ride.c [hunk 3]
```diff
@@ -756,12 +756,24 @@ int ride_find_track_gap(rct_xy_element *input, rct_xy_element *output)
...
 	while (track_circuit_iterator_next(&it)) {
 		if (!track_is_connected_by_shape(it.last.element, it.current.element)) {
 			*output = it.current;
 			return 1;
 		}
+		//#2081: prevent an infinite loop
+		counter = !counter;
+		if (counter) {
+			track_circuit_iterator_next(&slowIt);
+			if (slowIt.currentZ == it.currentZ && slowIt.currentDirection == it.currentDirection && slowIt.current.x == it.current.x && slowIt.current.y == it.current.y) {
```
Comment (duncanspumpkin) [inline src/ride/ride.c]: Put this on multiple lines it's far too hard to read as one long line

File: src/ride/ride.c [hunk 4]
```diff
@@ -756,12 +756,24 @@ int ride_find_track_gap(rct_xy_element *input, rct_xy_element *output)
 	if (w != NULL && _rideConstructionState != RIDE_CONSTRUCTION_STATE_0 && _currentRideIndex == rideIndex)
 		sub_6C9627();
 
+	bool counter = 1;
```
Comment (janisozaur) [inline src/ride/ride.c]: initialise with `true`

**Comment by jensj12** [comment]
It still hangs when there's a backwards infinite loop. I guess that's where my switchback coaster recreation made the game crash, so I'll try to fix that first before merging.

**Comment by jensj12** [comment]
The game still allows building backwards infinite loops and opening rides that have them, but they won't crash the game anymore.

**Pull request merged**